h/services/group.py

```diff
@@ -119,6 +122,57 @@ def groupids_created_by(self, user):
             g.pubid for g in self.session.query(Group.pubid).filter_by(creator=user)
         ]
 
+    @classmethod
+    def _group_filter(cls, query, pubids_or_groupids: List[str]):
+        """
+        Apply a filter to a query to efficiently limit it to certain groups.
+
+        This works for a mixed list of pubids or groupids.
+        """
+
+        # We will build up filter clauses for a single efficient query
+        clauses = []
+
+        if pubids := [
+            id_ for id_ in pubids_or_groupids if not group_util.is_groupid(id_)
+        ]:
+            clauses.append(cls._in_or_equals(Group.pubid, pubids))
+
+        if groupids := [
+            id_ for id_ in pubids_or_groupids if group_util.is_groupid(id_)
+        ]:
+            by_authority = defaultdict(list)
+            for parts in (group_util.split_groupid(groupid) for groupid in groupids):
+                by_authority[parts["authority"]].append(parts["authority_provided_id"])
+
+            # We'll probably only ever get called with a single authority, but
+            # there's no harm in being able to support multiple
+            for authority, authority_provided_ids in by_authority.items():
+                clauses.append(
+                    sa.and_(
+                        Group.authority == authority,
+                        cls._in_or_equals(
+                            Group.authority_provided_id, authority_provided_ids
+                        ),
+                    )
+                )
+
+        if len(clauses) == 1:
+            query = query.filter(clauses[0])
+        elif clauses:
+            query = query.filter(sa.or_(*clauses))
+
+        return query
```

Comment:
Kind of chungus, but this will build about the most efficient filter you can to search by both pubids and groupids at the same time.

This is in a separate method because:

 * It's huge and would disrupt the flow of the calling method
 * It's something we may want to apply to other queries in future (for example a `fetch_multiple_groups` would be trivial to write with this)